lib/ecto/parameterized_type.ex

```diff
@@ -0,0 +1,72 @@
+defmodule Ecto.ParameterizedType do
```

Comment:
I always thought we'd implement parameterised types as part of Ecto.Type. That is, instead of the current:

    @impl true
    def cast(term) do

folks would do:

    @impl true
    def cast(term, params)

and if your type is not parameterised you'd just define cast/1 and we'd define cast/2 for you as part of `use Ecto.Type`.

But perhaps it's better not to try shoehorning it in. Definitely better start to keep it as separate thing as you have, we always have an option to merge the concepts down the road if that would be an improvement.

This is only tangentially related, but I always considered our composite types (`{:array, _}` and `{:map, _}`) as another example of parameterised types besides embeds and we do handle them in Ecto.Type.

Reply:
@wojtekmach good to know, I will keep that in mind as I work on this. Yes, if we could pull it off, Parameterized Types being simply Ecto.Type but with opts would be nice. I would agree that at least for now, keeping it separate will allow us to concentrate on getting it done right, and we can merge down the road if possible.

And yes, I would love to eventually pull :array and :map out to parameterized, cleaning out a bunch of existing hardcoded code to handle them. Bonus points if we can figure out how to pull :assoc out to parameterized 😁.

Reply:
Unfortunately we cannot use Ecto.Type because of the "param" argument, so Ecto.ParameterizedType is the way to go IMO. If anything, we would make Ecto.Type become Ecto.ParameterizedType in the future (but most likely never).

Reply:
> And yes, I would love to eventually pull :array and :map out to parameterized, cleaning out a bunch of existing hardcoded code to handle them. Bonus points if we can figure out how to pull :assoc out to parameterized 😁.

This is actually more complex because the argument given to :array and :map is actually another type itself. To support them in `ParameterizedType`, then we would definitely need to pass a `dumper` function to `dump`, `loader` function to `load` and so on, because all of them are somewhat recursive with context. Maybe we should do that anyway or maybe embeds themselves already require this.